### PR TITLE
[WIP][HttpKernel] PSR6 Store for HttpCache

### PR DESCRIPTION
**Pull request by jakzal**
| Q | A |
| --- | --- |
| Branch? | master |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | - |
| License | MIT |
| Doc PR | not yet |

This is a WIP PR to get early feedback. I'd love to get this in for 3.2.

### CODE REVIEW

File: src/Symfony/Component/HttpKernel/HttpCache/Psr6Store.php [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+     * @param Request  $request  A Request instance
+     * @param Response $response A Response instance
+     *
+     * @return string The key under which the response is stored
+     */
+    public function write(Request $request, Response $response)
+    {
+        if (!$response->headers->has('X-Content-Digest')) {
+            $contentDigest = $this->generateContentDigest($response);
+            $item = $this->cacheBackend->getItem($contentDigest);
```
Comment (iltar) [inline src/Symfony/Component/HttpKernel/HttpCache/Psr6Store.php]: I would make this a guard clause instead. Please note that it should be `@return string|null`

File: src/Symfony/Component/HttpKernel/HttpCache/Psr6Store.php [hunk 2]
```diff
@@ -0,0 +1,197 @@
+<?php
+
+namespace Symfony\Component\HttpKernel\HttpCache;
+
+use Psr\Cache\CacheItemPoolInterface;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+
+class Psr6Store implements StoreInterface
+{
```
Comment (iltar) [inline src/Symfony/Component/HttpKernel/HttpCache/Psr6Store.php]: Any reason to not make it final?

Comment (jakzal) [inline src/Symfony/Component/HttpKernel/HttpCache/Psr6Store.php]: Nope. Good idea.

File: src/Symfony/Component/HttpKernel/HttpCache/Psr6Store.php (hunk 1 above)
Comment (jakzal) [inline src/Symfony/Component/HttpKernel/HttpCache/Psr6Store.php]: Actually, I should create a CacheItem instead. No need to query the backend as we're going to overwrite the value anyway (if it's there).

File: src/Symfony/Component/HttpKernel/HttpCache/Psr6Store.php [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+    /**
+     * @var CacheItemPoolInterface
+     */
+    private $cacheBackend;
```
Comment (ro0NL) [inline src/Symfony/Component/HttpKernel/HttpCache/Psr6Store.php]: What about `$cachePool` or even just `$pool`. See https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Cache/DoctrineProvider.php#L24

Comment (ro0NL) [inline src/Symfony/Component/HttpKernel/HttpCache/Psr6Store.php]: For being not in the `Cache` namespace, i'd favor `$cachePool`.

**Comment by iltar** [comment]
@jakzal this ticket is probably explaining the causation? #4871

**Comment by jakzal** [comment]
@iltar could be fixed by this indeed, but I was more thinking of making it more "small-website-prod-ready" (see #19705).

**Comment by iltar** [comment]
@jakzal yeah saw that issue as well when looking for the one mentioned. While this is not the best to use for production servers, it's a nice way to get started if you don't (yet) have a proper caching layer. I think this is a nice feature to have.

File: src/Symfony/Component/HttpKernel/HttpCache/Psr6Store.php (hunk 3 above)
Comment (jakzal) [inline src/Symfony/Component/HttpKernel/HttpCache/Psr6Store.php]: Changed for consistency :+1:

File: src/Symfony/Component/HttpKernel/HttpCache/Psr6Store.php [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+
+use Psr\Cache\CacheItemPoolInterface;
+use Symfony\Component\Cache\CacheItem;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+
+final class Psr6Store implements StoreInterface
+{
+    /**
+     * @var CacheItemPoolInterface
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/HttpCache/Psr6Store.php]: not needed, already hinted in the constructor

File: src/Symfony/Component/HttpKernel/HttpCache/Psr6Store.php [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+     */
+    private $cachePool;
+
+    /**
+     * List of locks acquired by the current process.
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/HttpCache/Psr6Store.php]: not sure this adds value

File: src/Symfony/Component/HttpKernel/HttpCache/Psr6Store.php [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+     *
+     * @var array
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/HttpCache/Psr6Store.php]: nor that

File: src/Symfony/Component/HttpKernel/HttpCache/Psr6Store.php [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+     */
+    private $locks = array();
+
+    /**
+     * @param CacheItemPoolInterface $cachePool
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/HttpCache/Psr6Store.php]: already in the constructor

File: src/Symfony/Component/HttpKernel/HttpCache/Psr6Store.php [hunk 8]
```diff
@@ -0,0 +1,242 @@
...
+     */
+    public function __construct(CacheItemPoolInterface $cachePool)
+    {
+        $this->cachePool = $cachePool;
+    }
+
+    /**
+     * Locates a cached Response for the Request provided.
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/HttpCache/Psr6Store.php]: `{@inheritdoc}`

File: src/Symfony/Component/HttpKernel/HttpCache/Psr6Store.php [hunk 9]
```diff
@@ -0,0 +1,242 @@
...
+     *
+     * @return Response|null A Response instance, or null if no cache entry was found
+     */
+    public function lookup(Request $request)
+    {
+        // TODO: Implement lookup() method.
+    }
+
+    /**
+     * Writes a cache entry to the store for the given Request and Response.
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/HttpCache/Psr6Store.php]: `{@inheritdoc}`

File: src/Symfony/Component/HttpKernel/HttpCache/Psr6Store.php [hunk 10]
```diff
@@ -0,0 +1,242 @@
...
+            if (!$response->headers->has('Transfer-Encoding')) {
+                $response->headers->set('Content-Length', strlen($response->getContent()));
+            }
+        }
+
+        $key = $this->getCacheKey($request);
+        $headers = $response->headers->all();
+        unset($headers['age']);
+
+        $this->save($key, serialize(array(array($request->headers->all(), $headers))));
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/HttpCache/Psr6Store.php]: serializing is already done by the cache pool

File: src/Symfony/Component/HttpKernel/HttpCache/Psr6Store.php [hunk 11]
```diff
@@ -0,0 +1,242 @@
...
+    }
+
+    /**
+     * Invalidates all cache entries that match the request.
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/HttpCache/Psr6Store.php]: `{@inheritdoc}`

Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/HttpCache/Psr6Store.php]: (same below :) )

File: src/Symfony/Component/HttpKernel/HttpCache/Psr6Store.php [hunk 12]
```diff
@@ -0,0 +1,242 @@
...
+     */
+    public function isLocked(Request $request)
+    {
+        $lockKey = $this->getLockKey($request);
+
+        if (isset($this->locks[$lockKey])) {
+            return true;
+        }
+
+        return $this->cachePool->hasItem($this->getLockKey($request));
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/HttpCache/Psr6Store.php]: use `$lockKey`

File: src/Symfony/Component/HttpKernel/HttpCache/Psr6Store.php [hunk 13]
```diff
@@ -0,0 +1,242 @@
...
+    }
+
+    /**
+     * @param string $key
+     * @param mixed  $value
+     * @param bool   $isHit
+     *
+     * @return CacheItem
+     */
+    private function createCacheItem($key, $value, $isHit = false)
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/HttpCache/Psr6Store.php]: This makes the implementation work only with Symfony's Cache components: cache items aren't interoperable.
You must get the item from the pool first (and maybe keep this as an optimization when Symfony Cache is used, another idea being to keep track on fetched items in this storage object to use them here)

**Review by nicolas-grekas — COMMENTED** [review]
Looks like a good start :) Would it make sense to make an AbstractStorage to share a few things with other storages implems?

File: src/Symfony/Component/HttpKernel/HttpCache/Psr6Store.php (hunk 13 above)
Comment (jakzal) [inline src/Symfony/Component/HttpKernel/HttpCache/Psr6Store.php]: This is kind of feedback i wanted to get (ignore comments etc, i left them there for my reference until I'm done and replace them with inheritdoc) ;)

> You must get the item from the pool first 

I did it first, perhaps I tried to optimise too early and replaced a call to pool with this. I'll revert back. Once it's all working we can run tests and see if it needs improving.

**Comment by jakzal** [comment]
> Would it make sense to make an AbstractStorage to share a few things with other storages implems?

I thought about it, but the current implementation is very coupled to the filesystem. I'm currently mirroring the behaviour, but the current implementation is optimised for the filesystem. I expect the psr6 implementation diverge eventually to be optimised for the cache pool.

 I'll try this approach nevertheless, to see which one makes more sense.

**Comment by fabpot** [comment]
Just for the record, I'm very skeptical about this one. We explicitly tied the HTTP cache to the filesystem.

**Comment by jakzal** [comment]
@fabpot no worries. If it's rejected here, I'll create a separate library (so work won't be wasted).

**Comment by Toflar** [comment]
Now that there's a `Lock` component, do things get easier? @jakzal do you plan to continue the work on this?

File: src/Symfony/Component/HttpKernel/HttpCache/Psr6Store.php (hunk 13 above)
Comment (stof) [inline src/Symfony/Component/HttpKernel/HttpCache/Psr6Store.php]: Having the optimization is OK if you keep the non-optimized version for other pools

**Comment by Toflar** [comment]
FYI: I added a PR to the fork of @jakzal to push this further: https://github.com/jakzal/symfony/pull/2
I was not sure whether I should just open a new PR because I actually like working together on something so I figured I'd just PR against the PR 😄

**Comment by Toflar** [comment]
@jakzal any feedback here? Even if it's just a "sorry, no time, please open a new PR"? 😄

**Comment by fabpot** [comment]
I'm closing this one as I think this is a bad idea for several reasons:

 * Using a filesystem allows for opcache to make the cache very effective;
 * The cache contains some PHP (when using ESI for instance) and storing PHP in anything else than a filesystem would mean eval()ing strings coming from Redis/Memcache/...,
 * HttpCache is triggered very early and does not have access to the container or anything else really. And it should stay that way to be efficient.

So, for these reasons, it won't be in Symfony core.

**Comment by Toflar** [comment]
Leaving it here for people looking for a solution without it being in Symfony core: https://packagist.org/packages/toflar/psr6-symfony-http-cache-store

**Comment by dunglas** [comment]
Relying on PSR-6 has a nice side-effect: the ability to use cache tags (see https://github.com/api-platform/core/pull/1689). Is supporting cache tags something we may want in core?

**Pull request closed without merging**